### PR TITLE
Reduce expansion of large types in the optimizer

### PR DESCRIPTION
**Pull request by shajrawi**
radar rdar://problem/33942706

This PR aims to reduce code size by limiting the explosions in the optimizer.

Note: Initial PR for benchmarking

### CODE REVIEW

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

File: lib/SIL/TypeLowering.cpp [hunk 1]
```diff
@@ -2520,3 +2520,69 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
 }
+
+size_t Lowering::countNumberOfFields(SILModule &Module, SILType Ty) {
+  size_t fieldsCount = 0;
+  CanType canType = Ty.getSwiftRValueType();
+
+  std::function<bool(Type)> traverseAndCount = [&](Type type) -> bool {
+    if (auto *structDecl = type->getStructOrBoundGenericStruct()) {
+      for (auto prop : structDecl->getStoredProperties()) {
+        if (!prop->hasType()) {
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: hasType() will always be false for deserialized properties

File: lib/SIL/TypeLowering.cpp [hunk 2]
```diff
@@ -2520,3 +2520,69 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+          continue;
+        }
+        auto propTy = prop->getType();
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: Use getInterfaceType() instead. At the SIL level you will only have fully type checked declarations, so it will never return null

File: lib/SIL/TypeLowering.cpp [hunk 3]
```diff
@@ -2520,3 +2520,69 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+        size_t fieldsCountBefore = fieldsCount;
+        propTy.findIf(traverseAndCount);
+        if (fieldsCount == fieldsCountBefore) {
+          // size of Struct(BigStructType) == sizeo of BigStructType()
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: Typo: "sizeo"

File: lib/SIL/TypeLowering.cpp [hunk 4]
```diff
@@ -2520,3 +2520,69 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+        if (fieldsCount > maxEnumCount) {
+          maxEnumCount = fieldsCount;
+        }
+      }
+      fieldsCount = fieldsCountBefore + maxEnumCount;
+    }
+    return true;
+  };
+
+  canType.findIf(traverseAndCount);
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: I don't think you want to use findIf() here -- this does a recursive traversal

File: lib/SIL/TypeLowering.cpp [hunk 5]
```diff
@@ -2520,3 +2520,69 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+    } else if (type->is<TupleType>()) {
+      auto tupleTy = type->castTo<TupleType>();
+      for (auto elt : tupleTy->getElementTypes()) {
+        auto canElt = elt->getCanonicalType();
+        canElt.findIf(traverseAndCount);
+      }
+    } else if (auto *enumDecl = type->getEnumOrBoundGenericEnum()) {
+      if (enumDecl->classifyAsOptionalType()) {
+        // Special case:
+        // The traverser will handle the optional payload on its own
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: Where is this handled?

File: lib/SIL/TypeLowering.cpp [hunk 6]
```diff
@@ -2520,3 +2520,69 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+          // size of Struct(BigStructType) == sizeo of BigStructType()
+          // prevent counting its size as BigStructType()+1
+          ++fieldsCount;
+        }
+      }
+    } else if (type->is<TupleType>()) {
+      auto tupleTy = type->castTo<TupleType>();
+      for (auto elt : tupleTy->getElementTypes()) {
+        auto canElt = elt->getCanonicalType();
+        canElt.findIf(traverseAndCount);
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: The outer findIf() will already recurse for you here

File: lib/SIL/TypeLowering.cpp [hunk 7]
```diff
@@ -2520,3 +2520,69 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+      size_t fieldsCountBefore = fieldsCount;
+      size_t maxEnumCount = 0;
+      for (auto elt : enumDecl->getAllElements()) {
+        if (!elt->hasAssociatedValues()) {
+          continue;
+        }
+        auto eltTy = currSILType.getEnumElementType(elt, Module);
+        auto eltCanTy = eltTy.getSwiftRValueType();
+        fieldsCount = 0;
+        eltCanTy.findIf(traverseAndCount);
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: This will recurse forever if you have an enum with an indirect case.

```
enum Foo {
  indirect case a(Foo)
  case b
}
```

File: lib/SIL/TypeLowering.cpp [hunk 8]
```diff
@@ -2520,3 +2520,69 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+          maxEnumCount = fieldsCount;
+        }
+      }
+      fieldsCount = fieldsCountBefore + maxEnumCount;
+    }
+    return true;
+  };
+
+  canType.findIf(traverseAndCount);
+  return std::max(fieldsCount, 1UL);
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: You might want to cache the result of this computation, since it's performed in a lot of places

**Review by slavapestov — CHANGES_REQUESTED** [review]
Maybe you want to use the substituted field type when walking into nominal declarations instead?

Eg, suppose I have

```
struct S<T> {
  let x: T
  let y: T
}

struct R {
  let a: S<(Int, Int), (Int, Int)>
}
```

Does `R` have 2 fields, or 4?

**Comment by slavapestov** [comment]
@shajrawi SIL type lowering already recursively walks all the fields of aggregates to figure out if a type is address only or loadable, etc.

It correctly computes substituted field types, supports indirect enums, and caching. Perhaps we should compute the 'size' as part of SIL type lowering, then you will just call `getTypeLowering(type).getRecursiveFieldCount()` or whatever.

Eventually we want SIL to stop looking at the AST for stored properties, because one day the optimizer will be able to remove or change stored properties of internal types. So adding new dependencies on the AST is undesirable here.

You can already see how we would have to encode knowledge of 'indirect' enum cases in your pass, which duplicates existing code for dealing with indirect cases.

**Comment by eeckstein** [comment]
> Eventually we want SIL to stop looking at the AST for stored properties, because one day the optimizer will be able to remove or change stored properties of internal types. So adding new dependencies on the AST is undesirable here.

@slavapestov So what would be the alternative for now?

**Comment by slavapestov** [comment]
@eeckstein I think a better approach is to calculate this 'recursive size' in type lowering and store it in the cached TypeLowering object. We already walk all fields anyway to determine if the type is address only or loadable.

**Comment by slavapestov** [comment]
@eeckstein Longer-term I would like SILGen to calculate 'lowered layout' objects for nominal types, and have the SIL optimizer and IRGen only look at those. The layout would order fields in physical order (not source order, if the two differ), explicitly reflect bit-packed fields (if we do that), make enum indirect cases explicit, etc.

**Comment by rjmccall** [comment]
I can see how optimizations that add and remove storage would benefit from that, but most of the other layout changes wouldn't.  Like why would SILGen benefit from putting fields in physical order or explicitly representing bit-packing?

**Comment by shajrawi** [comment]
@slavapestov I updated the PR with the changes we discussed off-line + caching of the results so that if we call `shouldExpand` on a Type that we already processed we'd return the cached result.

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@rjmccall SILGen would benefit from having fields in physical order because then we could conceivably change the packing algorithm without breaking the ABI.

File: include/swift/SIL/TypeLowering.h [hunk 9]
```diff
@@ -632,6 +634,10 @@ class TypeConverter {
     return ti.getLoweredType();
   }
 
+  ///  True if a type can be expanded
```
Comment (slavapestov) [inline include/swift/SIL/TypeLowering.h]: Perhaps shouldExpand() should live in the optimizer, and not in SIL itself?

File: include/swift/SIL/TypeLowering.h [hunk 10]
```diff
@@ -859,6 +865,8 @@ class TypeConverter {
                                AbstractionPattern pattern,
                                CanType result,
                                bool suppressOptional);
+
+  size_t countNumberOfFields(SILType);
```
Comment (slavapestov) [inline include/swift/SIL/TypeLowering.h]: Please add a comment here and make this method public.

File: include/swift/SIL/TypeLowering.h [hunk 11]
```diff
@@ -528,7 +528,9 @@ class TypeConverter {
   llvm::DenseMap<OverrideKey, SILConstantInfo> ConstantOverrideTypes;
 
   llvm::DenseMap<AnyFunctionRef, CaptureInfo> LoweredCaptures;
-  
+
+  llvm::DenseMap<SILType, size_t> TypeFields;
```
Comment (slavapestov) [inline include/swift/SIL/TypeLowering.h]: Everywhere else we use 'unsigned' to count the number of fields. I don't know if you'll want a struct with more than 4 billion members :-)

File: lib/SIL/TypeLowering.cpp [hunk 12]
```diff
@@ -2520,3 +2520,75 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
   
   return boxType;
 }
+
+static void countNumberOfInnerFields(size_t &fieldsCount, SILModule &Module,
+                                     SILType Ty) {
+  if (auto *structDecl = Ty.getStructOrBoundGenericStruct()) {
+    if (!structDecl->hasFixedLayout(Module.getSwiftModule(),
+                                    ResilienceExpansion::Minimal)) {
+      return;
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: Perhaps this should be an assertion instead, because FSO should not be trying to explode resilient (ie address-only) types at all.

File: lib/SIL/TypeLowering.cpp [hunk 13]
```diff
@@ -2520,3 +2520,75 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+      SILType propTy = Ty.getFieldType(prop, Module);
+      size_t fieldsCountBefore = fieldsCount;
+      countNumberOfInnerFields(fieldsCount, Module, propTy);
+      if (fieldsCount == fieldsCountBefore) {
+        // size of Struct(BigStructType) == size of BigStructType()
+        // prevent counting its size as BigStructType()+1
+        ++fieldsCount;
+      }
+    }
+  } else if (Ty.is<TupleType>()) {
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: `else if (auto tupleTy = Ty.getAs<TupleType>()) {`

File: lib/SIL/TypeLowering.cpp [hunk 14]
```diff
@@ -2520,3 +2520,75 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+    if (enumDecl->isIndirect()) {
+      return;
+    }
+    if (!enumDecl->hasFixedLayout(Module.getSwiftModule(),
+                                  ResilienceExpansion::Minimal)) {
+      return;
+    }
+    size_t fieldsCountBefore = fieldsCount;
+    size_t maxEnumCount = 0;
+    for (auto elt : enumDecl->getAllElements()) {
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: Actually don't you want to count an enum as 1 'field' always? After all, enums are never 'exploded' into scalars.

File: lib/SIL/TypeLowering.cpp [hunk 15]
```diff
@@ -2520,3 +2520,75 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+
+size_t TypeConverter::countNumberOfFields(SILType Ty) {
+  size_t fieldsCount = 0;
+  countNumberOfInnerFields(fieldsCount, M, Ty);
+  return std::max(fieldsCount, 1UL);
+}
+
+bool TypeConverter::shouldExpand(SILType Ty) {
+  size_t numFields = 0;
+  if (TypeFields.find(Ty) == TypeFields.end()) {
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: I would put the caching logic inside countNumberOfFields() instead of here.

File: lib/SILOptimizer/Utils/Local.cpp [hunk 16]
```diff
@@ -225,6 +225,9 @@ void swift::eraseUsesOfInstruction(SILInstruction *Inst,
   for (auto UI = Inst->use_begin(), E = Inst->use_end(); UI != E;) {
     auto *User = UI->getUser();
     UI++;
+    if (!User) {
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Local.cpp]: This looks like an unrelated fix, maybe it should be its own commit or PR?

**Review by slavapestov — COMMENTED** [review]
This looks good, I just have a few minor suggestions.

File: lib/SIL/TypeLowering.cpp [hunk 17]
```diff
@@ -2520,3 +2520,75 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+      auto silElt = SILType::getPrimitiveObjectType(elt->getCanonicalType());
+      countNumberOfInnerFields(fieldsCount, Module, silElt);
+    }
+  } else if (auto *enumDecl = Ty.getEnumOrBoundGenericEnum()) {
+    if (enumDecl->isIndirect()) {
+      return;
+    }
+    if (!enumDecl->hasFixedLayout(Module.getSwiftModule(),
+                                  ResilienceExpansion::Minimal)) {
+      return;
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: See above about resilient types

File: lib/SIL/TypeLowering.cpp [hunk 18]
```diff
@@ -2520,3 +2520,75 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+        maxEnumCount = fieldsCount;
+      }
+    }
+    fieldsCount = fieldsCountBefore + maxEnumCount;
+  }
+}
+
+size_t TypeConverter::countNumberOfFields(SILType Ty) {
+  size_t fieldsCount = 0;
+  countNumberOfInnerFields(fieldsCount, M, Ty);
```
Comment (slavapestov) [inline lib/SIL/TypeLowering.cpp]: I think countNumberOfInnerFields() should just return a count instead of modifying an out parameter by reference. Then you won't need a separate countNumberOfInnerFields() at all, TypeConverter::countNumberOfFields() could just be recursive.

**Comment by rjmccall** [comment]
@slavapestov I think we can achieve than better + more robustly without needing to involve SILGen.  Probably this isn't the best venue for this discussion, though. :)

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

**Comment by shajrawi** [comment]
@slavapestov I made the small changes in your review + made the heuristic more conservative to (hopefully) avoid some of the slowdowns - running benchmarks again

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

File: lib/SILOptimizer/Utils/Local.cpp [hunk 19]
```diff
@@ -2831,6 +2834,18 @@ bool swift::simplifyUsers(SILInstruction *I) {
   return Changed;
 }
 
+///  True if a type can be expanded
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Local.cpp]: Fixt he formatting here

File: lib/SILOptimizer/Utils/Local.cpp [hunk 20]
```diff
@@ -2831,6 +2834,18 @@ bool swift::simplifyUsers(SILInstruction *I) {
...
+/// without a significant increase to code size.
+bool swift::shouldExpand(SILModule &Module, SILType Ty) {
+  unsigned numFields = Module.Types.countNumberOfFields(Ty);
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Local.cpp]: why not just return (numFields > 6);

**Review by slavapestov — APPROVED** [review]
This looks good, can you clarify why you're considering enums at all though?

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

**Comment by shajrawi** [comment]
@swift-ci Please benchmark

File: lib/SILOptimizer/Transforms/SILLowerAggregateInstrs.cpp [hunk 21]
```diff
@@ -109,17 +112,23 @@ static bool expandCopyAddr(CopyAddrInst *CA) {
     //   retain_value %new : $*T
     IsTake_t IsTake = CA->isTakeOfSrc();
     if (IsTake_t::IsNotTake == IsTake) {
-      TL.emitLoweredCopyValue(Builder, CA->getLoc(), New,
-                              TypeLowering::LoweringStyle::Deep);
+      if (expand)
+        TL.emitLoweredCopyValue(Builder, CA->getLoc(), New,
+                                TypeLowering::LoweringStyle::Deep);
+      else
+        TL.emitCopyValue(Builder, CA->getLoc(), New);
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/SILLowerAggregateInstrs.cpp]: I think you can clean this up if you fix a bug in TypeLowering.h. In TypeLowering.cpp. The specific problem is that emitLoweredCopyValue ignores the LoweringStyle flag and always performs a deep copy. If you fix that so that it just checks the lowering and does a copy_value if it is shallow, you could clean up the multiple instances of this code by creating a helper function that goes from expand to LoweringStyle:

```
static TypeLowering::LoweringStyle getLoweringStyle(bool shouldExpand) {
    if (shouldExpand)
       return TypeLowering::LoweringStyle::Deep;
    return TypeLowering::LoweringStyle::Shallow;
}
```

Then you could just write this code and hide the conditional:

```
TL.emitLoweredCopyValue(Builder, CA->getLoc(), New, getLoweringStyle(expand));
```

IMO, the conditional does not really add anything to the flow of the code and is just noise.

File: lib/SILOptimizer/Transforms/SILLowerAggregateInstrs.cpp [hunk 22]
```diff
@@ -144,14 +153,19 @@ static bool expandDestroyAddr(DestroyAddrInst *DA) {
...
   // If we have a non-trivial type...
   if (!Type.isTrivial(Module)) {
     // If we have a type with reference semantics, emit a load/strong release.
     LoadInst *LI = Builder.createLoad(DA->getLoc(), Addr,
                                       LoadOwnershipQualifier::Unqualified);
     auto &TL = Module.getTypeLowering(Type);
-    TL.emitLoweredDestroyValue(Builder, DA->getLoc(), LI,
-                               TypeLowering::LoweringStyle::Deep);
+    if (expand)
+      TL.emitLoweredDestroyValue(Builder, DA->getLoc(), LI,
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/SILLowerAggregateInstrs.cpp]: This you could do a similar thing using perhaps a refactored helper method on LoweringStyle or TypeLowering. In this case, since emitLoweredDestroyValue is already correct, you would just get:

```
TL.emitLoweredDestroyValue(Builder, DA->GetLoc(), LI, getLoweringStyle(expand));
```

File: lib/SILOptimizer/Transforms/SILMem2Reg.cpp [hunk 23]
```diff
@@ -347,8 +347,13 @@ static void replaceDestroy(DestroyAddrInst *DAI, SILValue NewValue) {
 
   auto Ty = DAI->getOperand()->getType();
   auto &TL = DAI->getModule().getTypeLowering(Ty);
-  TL.emitLoweredDestroyValue(Builder, DAI->getLoc(), NewValue,
-                             Lowering::TypeLowering::LoweringStyle::Deep);
+
+  if (shouldExpand(DAI->getModule(),
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/SILMem2Reg.cpp]: Isn't this flipped? Also you could do the refactoring I mentioned here as well.

File: lib/SILOptimizer/Utils/Local.cpp [hunk 24]
```diff
@@ -2831,6 +2834,21 @@ bool swift::simplifyUsers(SILInstruction *I) {
...
 
+/// True if a type can be expanded
+/// without a significant increase to code size.
+bool swift::shouldExpand(SILModule &Module, SILType Ty) {
+  if (Ty.isAddressOnly(Module)) {
+    return false;
+  }
+  unsigned numFields = Module.Types.countNumberOfFields(Ty);
+  if (numFields > 6) {
+    // llvm::dbgs() << " should not expand " << Ty << " (" << numFields <<
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/Local.cpp]: You left in a dbgs statement here.

File: lib/SILOptimizer/Utils/Local.cpp [hunk 25]
```diff
@@ -2831,6 +2834,21 @@ bool swift::simplifyUsers(SILInstruction *I) {
   return Changed;
 }
 
+/// True if a type can be expanded
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/Local.cpp]: git-clang-format? = ).

File: test/SILOptimizer/lslocation_expansion.sil [hunk 26]
```diff
@@ -138,87 +138,9 @@ sil @store_after_store_struct : $@convention(thin) () -> () {
 // CHECK: #1 store
 // CHECK-NEXT: alloc_stack $S3
 // CHECK-NEXT: Projection Path [$*S3
-// CHECK-NEXT:   Field: var c: S2 of: $*S2
```
Comment (gottesmm) [inline test/SILOptimizer/lslocation_expansion.sil]: The problem with just deleting this code is that this code is making sure that this file is actually testing lslocation's ability to expand structs at all.

Can you add a flag to force expanding and use that here. Then you can (perhaps with a second RUN line/FileCheck) check it with expanding disabled (giving you what you are looking for, proof that you are not expanding).

**Review by gottesmm — COMMENTED** [review]
Some fly by comments.

File: lib/SILOptimizer/Transforms/SILLowerAggregateInstrs.cpp (hunk 21 above)
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/SILLowerAggregateInstrs.cpp]: Actually, LoweringStyle is a 2 case enum! So you could do LoweringStyle(shouldExpand). My only concern about that is if we add an additional LoweringStyle.

File: lib/SILOptimizer/Utils/Local.cpp (hunk 24 above)
Comment (shajrawi) [reply lib/SILOptimizer/Utils/Local.cpp]: whoops!

File: lib/SILOptimizer/Utils/Local.cpp (hunk 25 above)
Comment (shajrawi) [reply lib/SILOptimizer/Utils/Local.cpp]: did git-clang-format again - didn't complain about that one :)

File: lib/SILOptimizer/Transforms/SILLowerAggregateInstrs.cpp (hunk 21 above)
Comment (shajrawi) [reply lib/SILOptimizer/Transforms/SILLowerAggregateInstrs.cpp]: Done! looks much better

File: lib/SIL/TypeLowering.cpp [hunk 27]
```diff
@@ -2520,3 +2520,66 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+    for (auto prop : structDecl->getStoredProperties()) {
+      SILType propTy = Ty.getFieldType(prop, Module);
+      unsigned fieldsCountBefore = fieldsCount;
+      countNumberOfInnerFields(fieldsCount, Module, propTy);
+      if (fieldsCount == fieldsCountBefore) {
+        // size of Struct(BigStructType) == size of BigStructType()
+        // prevent counting its size as BigStructType()+1
+        ++fieldsCount;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (shajrawi) [reply lib/SIL/TypeLowering.cpp]: added early exit.

File: lib/SILOptimizer/Utils/Local.cpp (hunk 25 above)
Comment (gottesmm) [reply lib/SILOptimizer/Utils/Local.cpp]: Ok. My bad. ; ).

File: test/SILOptimizer/lslocation_expansion.sil (hunk 26 above)
Comment (shajrawi) [reply test/SILOptimizer/lslocation_expansion.sil]: We were discussing having multiple levels of expansions / a user-facing flag (maybe depending on the optimization / code-size level) - we will add that in a later date which would also cover this test case.

Comment (gottesmm) [reply test/SILOptimizer/lslocation_expansion.sil]: How soon are you talking about doing that work?

I would really really not like to lose this code coverage. How about just adding an LLVM option at expand that just causes it to always return true. It is not user exposed and just for testing. Then copy/paste the FileCheck tests with a different prefix.

File: include/swift/SIL/TypeLowering.h [hunk 28]
```diff
@@ -528,7 +536,9 @@ class TypeConverter {
   llvm::DenseMap<OverrideKey, SILConstantInfo> ConstantOverrideTypes;
 
   llvm::DenseMap<AnyFunctionRef, CaptureInfo> LoweredCaptures;
-  
+
+  llvm::DenseMap<SILType, unsigned> TypeFields;
```
Comment (gottesmm) [inline include/swift/SIL/TypeLowering.h]: Can you add a comment here?

File: include/swift/SIL/TypeLowering.h [hunk 29]
```diff
@@ -575,7 +585,10 @@ class TypeConverter {
   
   /// Get the method dispatch strategy for a protocol.
   static ProtocolDispatchStrategy getProtocolDispatchStrategy(ProtocolDecl *P);
-  
+
+  /// Count the total number of fields inside the given SIL Type
+  unsigned countNumberOfFields(SILType);
```
Comment (gottesmm) [inline include/swift/SIL/TypeLowering.h]: Can you add a variable name here just so we are consistent in the file?

File: lib/SIL/TypeLowering.cpp (hunk 27 above)
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SIL/TypeLowering.cpp]: That isn't what I meant. I meant this:

```
if (...) {
   *code*...
   return...
}
<=== *No else*

if (...) {
   ... *code* ...
   return.
}
<=== *No else *

*rest of code*
```

That is the canonical LLVM way of writing this pattern.

File: lib/SIL/TypeLowering.cpp [hunk 30]
```diff
@@ -2520,3 +2520,68 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+        maxEnumCount = fieldsCount;
+      }
+    }
+    fieldsCount = fieldsCountBefore + maxEnumCount;
+  }
+}
+
+unsigned TypeConverter::countNumberOfFields(SILType Ty) {
+  unsigned fieldsCount = 0;
+  if (TypeFields.find(Ty) == TypeFields.end()) {
```
Comment (gottesmm) [inline lib/SIL/TypeLowering.cpp]: Can you not recompute here and return early? You are performing two hash table lookups. You do not need to do this. TypeFields.find will return an iterator that will have fieldsCount in it, i.e.:

```
auto Iter = TypeFields.find(Ty);
if (Iter == TypeFields.end()) {
    unsigned fieldsCount = 0;
    countNumberOfInnerFields(fieldsCount, M, Ty);
    TypeFields[Ty] = fieldsCount;
    return std::max(fieldsCount, 1U);
}

return std::max(Iter->second, 1U);
```

Or actually, it would be even cleaner if you flipped the control flow.
```
auto Iter == TypeFields.find(Ty);
if (Iter != TypeFields.end()) {
    return std::max(Iter->second, 1U);
}

unsigned fieldsCount = 0;
countNumberOfInnerFields(fieldsCount, M, Ty);
TypeFields[Ty] = fieldsCount;
return std::max(fieldsCount, 1U);
```

File: lib/SILOptimizer/Transforms/SILLowerAggregateInstrs.cpp [hunk 31]
```diff
@@ -110,7 +113,7 @@ static bool expandCopyAddr(CopyAddrInst *CA) {
     IsTake_t IsTake = CA->isTakeOfSrc();
     if (IsTake_t::IsNotTake == IsTake) {
       TL.emitLoweredCopyValue(Builder, CA->getLoc(), New,
-                              TypeLowering::LoweringStyle::Deep);
+                              TypeLowering::getLoweringStyle(expand));
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/SILLowerAggregateInstrs.cpp]: This and the rest of this file looks awesome! Nice clean and easy to read! +1!

File: lib/SILOptimizer/Utils/Local.cpp (hunk 16 above)
Comment (gottesmm) [reply lib/SILOptimizer/Utils/Local.cpp]: At least add a comment here to explain what is going on with the iterator. If this is correct, we are relying on some sort of special semantics to make sure we are memory safe.

**Review by gottesmm — COMMENTED** [review]
Some more comments.

File: lib/SIL/TypeLowering.cpp [hunk 32]
```diff
@@ -2520,3 +2524,68 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
 }
+
+static void countNumberOfInnerFields(unsigned &fieldsCount, SILModule &Module,
+                                     SILType Ty) {
+  if (auto *structDecl = Ty.getStructOrBoundGenericStruct()) {
+    assert(structDecl->hasFixedLayout(Module.getSwiftModule(),
+                                      ResilienceExpansion::Minimal) &&
+           " FSO should not be trying to explode resilient (ie address-only) "
+           "types at all");
+    for (auto prop : structDecl->getStoredProperties()) {
```
Comment (gottesmm) [inline lib/SIL/TypeLowering.cpp]: Is this a pointer? Since this is a for loop, can you put in a * to match LLVM style.

File: lib/SIL/TypeLowering.cpp [hunk 33]
```diff
@@ -2520,3 +2524,68 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+      countNumberOfInnerFields(fieldsCount, Module, propTy);
+      if (fieldsCount == fieldsCountBefore) {
+        // size of Struct(BigStructType) == size of BigStructType()
+        // prevent counting its size as BigStructType()+1
+        ++fieldsCount;
+      }
+    }
+    return;
+  } else if (auto tupleTy = Ty.getAs<TupleType>()) {
+    for (auto elt : tupleTy->getElementTypes()) {
```
Comment (gottesmm) [inline lib/SIL/TypeLowering.cpp]: Same here. Can you make this an auto * if it is a pointer.

File: lib/SIL/TypeLowering.cpp [hunk 34]
```diff
@@ -2520,3 +2524,68 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+    unsigned fieldsCountBefore = fieldsCount;
+    unsigned maxEnumCount = 0;
+    for (auto elt : enumDecl->getAllElements()) {
+      if (!elt->getArgumentInterfaceType()) {
+        continue;
+      }
+      if (elt->isIndirect()) {
+        continue;
+      }
+      // Number of fields > 1 as "future proof" for this heuristic:
```
Comment (gottesmm) [inline lib/SIL/TypeLowering.cpp]: What is this comment talking about? I don't understand it without needing to think.

**Review by gottesmm — COMMENTED** [review]
Nits + an ask to expand/explain a comment.

File: lib/SIL/TypeLowering.cpp (hunk 27 above)
[the remark this answers is not in the document]
Comment (shajrawi) [reply lib/SIL/TypeLowering.cpp]: Done

File: test/SILOptimizer/lslocation_expansion.sil (hunk 26 above)
Comment (shajrawi) [reply test/SILOptimizer/lslocation_expansion.sil]: Alright I added a frontend option and turned it on for those tests.

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

File: include/swift/AST/SILOptions.h [hunk 35]
```diff
@@ -146,6 +146,9 @@ class SILOptions {
   /// \brief Enable large loadable types IRGen pass.
   bool EnableLargeLoadableTypes = true;
 
+  /// \brief Enable expansion of types regardless of size.
+  bool EnableExpandAll = false;
```
Comment (gottesmm) [inline include/swift/AST/SILOptions.h]: Ok. I was just imagining an LLVM option that could be put where-ever you defined expand. But this works as well.

File: lib/SIL/TypeLowering.cpp (hunk 27 above)
[the remark this answers is not in the document]
Comment (gottesmm) [reply lib/SIL/TypeLowering.cpp]: Looking good here!

File: lib/SIL/TypeLowering.cpp [hunk 36]
```diff
@@ -2520,3 +2524,74 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+    return;
+  }
+  if (auto tupleTy = Ty.getAs<TupleType>()) {
+    for (auto elt : tupleTy->getElementTypes()) {
+      auto silElt = SILType::getPrimitiveObjectType(elt->getCanonicalType());
+      countNumberOfInnerFields(fieldsCount, Module, silElt);
+    }
+    return;
+  }
+  if (auto *enumDecl = Ty.getEnumOrBoundGenericEnum()) {
```
Comment (gottesmm) [inline lib/SIL/TypeLowering.cpp]: You can reduce the indentation here further by doing:

```
auto *enumDecl = Ty.getEnumOrBoundGenericEnum();
if (!enumDecl)
  return;
```

File: lib/SIL/TypeLowering.cpp [hunk 37]
```diff
@@ -2520,3 +2524,74 @@ TypeConverter::getContextBoxTypeForCapture(ValueDecl *captured,
...
+        maxEnumCount = fieldsCount;
+      }
+    }
+    fieldsCount = fieldsCountBefore + maxEnumCount;
+    return;
+  }
+}
+
+unsigned TypeConverter::countNumberOfFields(SILType Ty) {
+  auto Iter = TypeFields.find(Ty);
```
Comment (gottesmm) [inline lib/SIL/TypeLowering.cpp]: Looking good!

File: include/swift/SIL/TypeLowering.h [hunk 38]
```diff
@@ -528,7 +536,10 @@ class TypeConverter {
   llvm::DenseMap<OverrideKey, SILConstantInfo> ConstantOverrideTypes;
 
   llvm::DenseMap<AnyFunctionRef, CaptureInfo> LoweredCaptures;
-  
+
+  // Cache of loadable SILType to number of (estimated) fields
```
Comment (gottesmm) [inline include/swift/SIL/TypeLowering.h]: This should be 3 slashes for doxygen purpose.

**Review by gottesmm — COMMENTED** [review]
1. A small comment fix.
2. 1 more way of reducing the indentation even more.

**Comment by gottesmm** [comment]
This is looking really good!

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test
